From 06e049af6fa2b5fefcdee38fa8db5204f9c59637 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Fri, 7 Mar 2025 09:22:18 +0100
Subject: [PATCH 1658/1695] prod/tfm-stm32mp: remove mutex in mailbox module

Since the scmi request are executed in a spm partition from tfm, the
execution can not be multithread, mutex is useless

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6a898a0c87a6836c04ac3305b5babfb4a4ec04de prod/tfm-stm32mp: remove mutex in mailbox module]
Change-Id: I13aa754c737e90b8d593eccc3bc51c14c76aab64
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/SCP-firmware/+/445563
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Valentin CARON <valentin.caron@foss.st.com>
Domain-Review: Valentin CARON <valentin.caron@foss.st.com>
---
 .../module/tfm_mbx/src/mod_tfm_mbx.c          | 52 -------------------
 1 file changed, 52 deletions(-)

diff --git a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c
index f7c08227c..38d2ea534 100644
--- a/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c
+++ b/core/lib/scmi-server/SCP-firmware/product/tfm-stm32mp2/module/tfm_mbx/src/mod_tfm_mbx.c
@@ -21,40 +21,6 @@
 #include <mod_tfm_mbx.h>
 #include <mod_tfm_smt.h>
 #include <mod_msg_smt.h>
-#include <cmsis_compiler.h>
-#define COMPILER_BARRIER() __ASM volatile("" : : : "memory")
-
-struct mutex {
-    uint8_t atomic;
-};
-
-static uint8_t atomic_add_u8(volatile uint8_t *valuePtr, int8_t delta)
-{
-    COMPILER_BARRIER();
-    uint8_t newValue;
-    do
-    {
-        newValue = __LDREXB(valuePtr) + delta;
-    } while (__STREXB(newValue, valuePtr));
-    COMPILER_BARRIER();
-    return newValue;
-}
-static void mutex_init(struct mutex *lock)
-{
-    lock->atomic = 0;
-}
-
-static void mutex_lock(struct mutex *lock)
-{
-    atomic_add_u8(&lock->atomic,1);
-    assert(lock->atomic == 1);
-}
-
-static void mutex_unlock(struct mutex *lock)
-{
-    atomic_add_u8(&lock->atomic,-1);
-    assert(lock->atomic == 0);
-}
 
 /* MBX device context */
 struct mbx_device_ctx {
@@ -69,8 +35,6 @@ struct mbx_device_ctx {
     } shmem_api;
 
     size_t *shm_out_size;
-
-    struct mutex lock;
 };
 
 /* MBX context */
@@ -98,10 +62,6 @@ void tfm_mbx_signal_msg_message(fwk_id_t device_id, void *in_buf,
         fwk_assert(fwk_id_get_module_idx(device_ctx->shmem_id) ==
                    FWK_MODULE_IDX_MSG_SMT);
 
-
-        /* Lock the channel until the message has been processed */
-        mutex_lock(&device_ctx->lock);
-
         device_ctx->shm_out_size = out_size;
         device_ctx->shmem_api.msg->signal_message(device_ctx->shmem_id,
                                                   in_buf, in_size,
@@ -140,9 +100,6 @@ void tfm_mbx_signal_smt_message(fwk_id_t device_id)
         fwk_assert(fwk_id_get_module_idx(device_ctx->shmem_id) ==
                    FWK_MODULE_IDX_TFM_SMT);
 
-        /* Lock the channel until the message has been processed */
-        mutex_lock(&device_ctx->lock);
-
         device_ctx->shmem_api.smt->signal_message(device_ctx->shmem_id);
     } else {
         fwk_unexpected();
@@ -156,9 +113,6 @@ static int raise_smt_interrupt(fwk_id_t channel_id)
     size_t idx = fwk_id_get_element_idx(channel_id);
     struct mbx_device_ctx *channel_ctx = &mbx_ctx.device_ctx_table[idx];
 
-    /* Release the channel as the message has been processed */
-    mutex_unlock(&channel_ctx->lock);
-
     /* There should be a message in the mailbox */
     return FWK_SUCCESS;
 }
@@ -177,10 +131,6 @@ static int raise_shm_notification(fwk_id_t channel_id, size_t size)
 
     *channel_ctx->shm_out_size = size;
 
-
-    /* Release the channel as the message has been processed */
-    mutex_unlock(&channel_ctx->lock);
-
     return FWK_SUCCESS;
 }
 
@@ -218,8 +168,6 @@ static int mbx_device_init(fwk_id_t device_id, unsigned int slot_count,
 
     device_ctx->config = (struct mod_tfm_mbx_channel_config*)data;
 
-    mutex_init(&device_ctx->lock);
-
     return FWK_SUCCESS;
 }
 
-- 
2.47.2

